test/client.js

```diff
@@ -108,6 +110,9 @@ describe('MqttClient', function () {
 
       client.on('message', function (t, p, packet) {
         if (++count === max) {
+          // BUGBUG: the client.end callback never gets called here
+          // client.end((err) => done(err))
```

Comment:
I use mqttjs in lot of my projects and this happend to me too, I always create a MqttClient wrapper library that handles all client stuff, this is how I handle the close usually: https://github.com/zwave-js/zwave-js-ui/blob/master/lib/MqttClient.ts#L154. I think this could happen when for some reason underlig socket is already closed/destroyed or when store cannot be closed for some reason

Reply:
No clue if this could help: https://github.com/mqttjs/MQTT.js/pull/713

Reply:
I merged #713 into a local branch and it didn't resolve this issue. It is an interesting change though. 

One of my goals with these test changes is to get the behavior of `client.end` well tested before we make any changes to it. Right now, it's just too hard to reason about how `client.end` might behave and I'm afraid to make any changes to it.

Reply:
I agree! Tests are essentials if we want to make improvements to the code